app/gui/src/searcher.rs

```diff
@@ -0,0 +1,196 @@
+use crate::presenter;
+use crate::presenter::graph::AstNodeId;
+use crate::presenter::graph::ViewNodeId;
+
+use ide_view as view;
+use ide_view::component_browser::component_list_panel::grid as component_grid;
+use ide_view::graph_editor::GraphEditor;
+use ide_view::graph_editor::NodeId;
+use ide_view::project::SearcherParams;
+use ide_view::project::SearcherType;
+
+
+
+/// Trait for the searcher.
+pub trait SearcherPresenter: Debug {
+    /// Initiate the operating mode for the searcher based on the given [`SearcherParams`]. If the
+    /// view associated with the input node given in the parameters does not yet exist, it will
+    /// be created.
+    ///
+    /// Returns the [`Mode`] that should be used for the searcher.
+    fn init_input_node(
+        parameters: SearcherParams,
+        graph: &presenter::Graph,
+        graph_editor: &GraphEditor,
+        graph_controller: &controller::Graph,
+    ) -> FallibleResult<Mode>
+    where
+        Self: Sized,
+    {
+        let SearcherParams { input, .. } = parameters;
+        let ast_node = graph.ast_node_of_view(input);
+
+        let mode: FallibleResult<_> = match ast_node {
+            Some(node_id) => Ok(Mode::EditNode { node_id }),
+            None => {
+                let (new_node, source_node) =
+                    create_input_node(parameters, graph, graph_editor, graph_controller)?;
+                Ok(Mode::NewNode { node_id: new_node, source_node })
+            }
+        };
+        let mode = mode?;
+        let target_node = mode.node_id();
+
+        if let Some(target_node_view) = graph.view_id_of_ast_node(target_node) {
+            // We only want to show the preview of the node if it is a component browser searcher.
+            if matches!(parameters.searcher_type, SearcherType::ComponentBrowser) {
+                graph_editor.model.with_node(target_node_view, |node| node.show_preview());
+            }
+        } else {
+            warn!("No view associated with node {:?}.", target_node);
+        }
+        // We disable auto-updates for the expression of the node, so we can set the expression
+        // of the input node without triggering an update of the graph. This is used, for example,
+        // to show a preview of the item selected in the component browser without changing the
+        // text the user has typed on the searcher input node.
+        graph.allow_expression_auto_updates(target_node, false);
+
+        Ok(mode)
+    }
+
+    /// Setup new, appropriate searcher controller for the edition of `node_view`, and construct
+    /// presenter handling it.
+    fn setup_searcher(
+        ide_controller: controller::Ide,
+        project_controller: controller::Project,
+        graph_controller: controller::ExecutedGraph,
+        graph_presenter: &presenter::Graph,
+        view: view::project::View,
+        parameters: SearcherParams,
+    ) -> FallibleResult<Box<dyn SearcherPresenter>>
+    where
+        Self: Sized;
```

Comment:
`Self` is not allowed in this place. Putting it there results in the following error:
```
pub trait SearcherPresenter: Debug {
   |           ----------------- `Self` not allowed in a trait
...
81 |     ) -> FallibleResult<Box<dyn Self>>
   |                                 ^^^^ `Self` is only available in impls, traits, and type definitions
```
 https://doc.rust-lang.org/error_codes/E0411.html

Reply:
I meant, you can just return Self, unboxed and without `dyn`.

Reply:
Ah. Yes, we could implement that, but in `setup_searcher_presenter` we need a `Box<dyn SearcherPresenter>` and this way all the methods are the same and can be processed nicer. Instead of having a separate call with all arguments for each searcher type, we just choose the needed constructor.

See https://github.com/enso-org/enso/blob/f2922ed8a6e6dcec1bcef3eefc989661901a5e83/app/gui/src/presenter/project.rs#L82

Reply:
My proposition is then to make `setup_searcher` return self and have a provided function in this trait `setup_searcher_boxed` which boxes `setup_searcher` and returns `FallibleResult<Box<dyn SearcherPresenter>>`

Reply:
Agree. Done.